src/ptr_array.rs

```diff
@@ -0,0 +1,218 @@
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use glib_sys;
+use std::fmt;
+use translate::*;
+
+glib_wrapper! {
+    pub struct PtrArray(Shared<glib_sys::GPtrArray>);
+
+    match fn {
+        ref => |ptr| glib_sys::g_ptr_array_ref(ptr),
+        unref => |ptr| glib_sys::g_ptr_array_unref(ptr),
+        get_type => || glib_sys::g_ptr_array_get_type(),
+    }
+}
+
+impl<'a, T> ToGlibContainerFromSlice<'a, *mut glib_sys::GPtrArray> for T
+where
+    T: GlibPtrDefault + ToGlibPtr<'a, <T as GlibPtrDefault>::GlibType>,
+{
+    type Storage = (
+        Option<PtrArray>,
+        Vec<Stash<'a, <T as GlibPtrDefault>::GlibType, T>>,
+    );
+
+    #[inline]
+    fn to_glib_none_from_slice(t: &'a [T]) -> (*mut glib_sys::GPtrArray, Self::Storage) {
+        let (ptr, storage) = Self::to_glib_container_from_slice(t);
+
+        unsafe { (ptr, (Some(from_glib_full(ptr)), storage.1)) }
+    }
+
+    #[inline]
+    fn to_glib_container_from_slice(t: &'a [T]) -> (*mut glib_sys::GPtrArray, Self::Storage) {
+        let stash_vec: Vec<_> = t.iter().map(ToGlibPtr::to_glib_none).collect();
+        unsafe {
+            let arr = glib_sys::g_ptr_array_new_full(t.len() as _, None);
+
+            for stash in &stash_vec {
+                let ptr: *mut <T as GlibPtrDefault>::GlibType = Ptr::to(stash.0);
+                glib_sys::g_ptr_array_add(arr, ptr as *mut _);
+            }
+
+            (arr, (None, stash_vec))
+        }
+    }
+
+    #[inline]
+    fn to_glib_full_from_slice(_t: &[T]) -> *mut glib_sys::GPtrArray {
+        unimplemented!()
+    }
+}
+
+impl<'a, T> ToGlibContainerFromSlice<'a, *const glib_sys::GPtrArray> for T
+where
+    T: GlibPtrDefault + ToGlibPtr<'a, <T as GlibPtrDefault>::GlibType>,
+{
+    type Storage = (
+        Option<PtrArray>,
+        Vec<Stash<'a, <T as GlibPtrDefault>::GlibType, T>>,
+    );
+
+    #[inline]
+    fn to_glib_none_from_slice(t: &'a [T]) -> (*const glib_sys::GPtrArray, Self::Storage) {
+        let (arr, stash) =
+            ToGlibContainerFromSlice::<*mut glib_sys::GPtrArray>::to_glib_none_from_slice(t);
+        (arr as *const glib_sys::GPtrArray, stash)
+    }
+
+    #[inline]
+    fn to_glib_container_from_slice(_t: &'a [T]) -> (*const glib_sys::GPtrArray, Self::Storage) {
+        unimplemented!()
+    }
+
+    #[inline]
+    fn to_glib_full_from_slice(_t: &[T]) -> *const glib_sys::GPtrArray {
+        unimplemented!()
+    }
+}
+
+impl<T> FromGlibContainerAsVec<<T as GlibPtrDefault>::GlibType, *mut glib_sys::GPtrArray> for T
+where
+    T: GlibPtrDefault
+        + FromGlibPtrNone<<T as GlibPtrDefault>::GlibType>
+        + FromGlibPtrFull<<T as GlibPtrDefault>::GlibType>,
+{
+    unsafe fn from_glib_none_num_as_vec(ptr: *mut glib_sys::GPtrArray, num: usize) -> Vec<T> {
+        if num == 0 || ptr.is_null() {
+            return Vec::new();
+        }
+        let pdata = (*ptr).pdata;
+        let mut res = Vec::with_capacity(num);
```

Comment:
Should first of all check here that the pointer array has `num` items available at all. Should probably become an assertion

Reply:
Added asserts at the two places I am using `with_capacity`.